### PR TITLE
support GET verb in AbstractClientStream2

### PR DESCRIPTION
**Pull request by zsurocking**
Based on ejona86@89f0d8c. It adds a writeHeaders API to AbstractClientStream2.Sink. When sending a GET request, writeHeaders call will be delayed until we have the request payload and halfClose is called. The request payload can then be sent in the query string.

### CODE REVIEW

File: netty/src/main/java/io/grpc/netty/NettyClientStream.java [hunk 1]
```diff
@@ -103,47 +101,55 @@ protected Sink abstractClientStreamSink() {
...
+      boolean get = (requestPayload != null);
+      if (get) {
+        // Forge the query string
+        defaultPath = new AsciiString(defaultPath + "?" + GRPC_PAYLOAD_BIN_KEY + "="
+            + BaseEncoding.base64().encode(requestPayload));
       }
-    };
+      Http2Headers http2Headers
+          = Utils.convertClientHeaders(headers, scheme, defaultPath, authority, userAgent,
+              get ? AsciiString.of("GET") : Utils.HTTP_METHOD);
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/NettyClientStream.java]: Make the GET asciistring a constant.

File: netty/src/test/java/io/grpc/netty/NettyClientStreamTest.java [hunk 2]
```diff
@@ -392,6 +393,38 @@ public void removeUserAgentFromApplicationHeaders() {
         .containsEntry(Utils.USER_AGENT, AsciiString.of("good agent"));
   }
 
+  @Test
+  public void getRequestSentThroughHeader() {
```
Comment (ejona86) [inline netty/src/test/java/io/grpc/netty/NettyClientStreamTest.java]: A test like this should be in AbstractClientStream2Test.java as well. At that point, this one could possibly become smaller.

File: netty/src/main/java/io/grpc/netty/NettyClientStream.java [hunk 3]
```diff
@@ -80,11 +78,11 @@
       TransportState state, MethodDescriptor<?, ?> method, Metadata headers,
       Channel channel, AsciiString authority, AsciiString scheme, AsciiString userAgent,
       StatsTraceContext statsTraceCtx) {
-    super(new NettyWritableBufferAllocator(channel.alloc()), statsTraceCtx);
+    super(new NettyWritableBufferAllocator(channel.alloc()), statsTraceCtx, headers,
```
Comment (carl-mastrangelo) [inline netty/src/main/java/io/grpc/netty/NettyClientStream.java]: wrap line to keep arguments on the same line.   That or do an argument per line.

File: netty/src/main/java/io/grpc/netty/NettyClientStream.java [hunk 4]
```diff
@@ -103,47 +101,55 @@ protected Sink abstractClientStreamSink() {
...
-    if (defaultPath == null) {
-      defaultPath = new AsciiString("/" + method.getFullMethodName());
-      methodDescriptorAccessor.setRawMethodName(method, defaultPath);
-    }
-    headers.discardAll(GrpcUtil.USER_AGENT_KEY);
-    Http2Headers http2Headers
-        = Utils.convertClientHeaders(headers, scheme, defaultPath, authority, userAgent);
-    headers = null;
+  private static boolean useGet(MethodDescriptor<?, ?> method) {
+    return method.isIdempotent() && method.isSafe()
```
Comment (carl-mastrangelo) [inline netty/src/main/java/io/grpc/netty/NettyClientStream.java]: Seems redundant to call both idempotent and safe.  Safe implies idempotent.

File: netty/src/main/java/io/grpc/netty/NettyClientStream.java [hunk 5]
```diff
@@ -103,47 +101,55 @@ protected Sink abstractClientStreamSink() {
...
+        && (method.getType() == MethodDescriptor.MethodType.UNARY);
```
Comment (carl-mastrangelo) [inline netty/src/main/java/io/grpc/netty/NettyClientStream.java]: No need to check if the type is unary.  MethodDescriptor enforces this invariant.

File: netty/src/main/java/io/grpc/netty/NettyClientStream.java (hunk 1 above)
Comment (carl-mastrangelo) [inline netty/src/main/java/io/grpc/netty/NettyClientStream.java]: Could you group the method header with the other pseudo headers?  Preferably before userAgent.

File: netty/src/main/java/io/grpc/netty/NettyClientStream.java [hunk 6]
```diff
@@ -103,47 +101,55 @@ protected Sink abstractClientStreamSink() {
...
+      // Convert the headers into Netty HTTP/2 headers.
+      AsciiString defaultPath = (AsciiString) methodDescriptorAccessor.geRawMethodName(method);
+      if (defaultPath == null) {
+        defaultPath = new AsciiString("/" + method.getFullMethodName());
+        methodDescriptorAccessor.setRawMethodName(method, defaultPath);
+      }
+      headers.discardAll(GrpcUtil.USER_AGENT_KEY);
+      boolean get = (requestPayload != null);
+      if (get) {
+        // Forge the query string
```
Comment (carl-mastrangelo) [inline netty/src/main/java/io/grpc/netty/NettyClientStream.java]: Optional: consider moving the assignment of the method along with assignment of the path

File: netty/src/main/java/io/grpc/netty/NettyClientStream.java [hunk 7]
```diff
@@ -103,47 +101,55 @@ protected Sink abstractClientStreamSink() {
...
-    checkState(headers != null, "must be call before start");
     this.authority = AsciiString.of(checkNotNull(authority, "authority"));
   }
 
   @Override
-  public void start(ClientStreamListener listener) {
-    super.start(listener);
+  public Attributes getAttributes() {
+    return state.handler.getAttributes();
+  }
```
Comment (carl-mastrangelo) [inline netty/src/main/java/io/grpc/netty/NettyClientStream.java]: nit: moving this made the diff harder to read :(

File: core/src/main/java/io/grpc/internal/AbstractClientStream2.java [hunk 8]
```diff
@@ -308,4 +333,70 @@ private void closeListener(Status status, Metadata trailers) {
...
+    @Override
+    public boolean isClosed() {
+      return closed;
+    }
+
+    /** Closes, with flush. */
+    @Override
+    public void close() {
+      closed = true;
+      if (payload == null) {
```
Comment (zsurocking) [inline core/src/main/java/io/grpc/internal/AbstractClientStream2.java]: done

File: netty/src/main/java/io/grpc/netty/NettyClientStream.java (hunk 1 above)
Comment (zsurocking) [inline netty/src/main/java/io/grpc/netty/NettyClientStream.java]: Added to Utils

File: netty/src/test/java/io/grpc/netty/NettyClientStreamTest.java (hunk 2 above)
Comment (zsurocking) [inline netty/src/test/java/io/grpc/netty/NettyClientStreamTest.java]: Added a test in AbstractClientStream2Test.java. I kept this one the same though.

File: netty/src/main/java/io/grpc/netty/NettyClientStream.java (hunk 3 above)
Comment (zsurocking) [inline netty/src/main/java/io/grpc/netty/NettyClientStream.java]: Done

File: netty/src/main/java/io/grpc/netty/NettyClientStream.java (hunk 7 above)
Comment (zsurocking) [inline netty/src/main/java/io/grpc/netty/NettyClientStream.java]: Not sure if anything I can do here... Do you want me to move this after Sink?

File: netty/src/main/java/io/grpc/netty/NettyClientStream.java (hunk 4 above)
Comment (zsurocking) [inline netty/src/main/java/io/grpc/netty/NettyClientStream.java]: Done

File: netty/src/main/java/io/grpc/netty/NettyClientStream.java (hunk 5 above)
Comment (zsurocking) [inline netty/src/main/java/io/grpc/netty/NettyClientStream.java]: Done

File: netty/src/main/java/io/grpc/netty/NettyClientStream.java (hunk 6 above)
Comment (zsurocking) [inline netty/src/main/java/io/grpc/netty/NettyClientStream.java]: Moved before headers.discardAll(GrpcUtil.USER_AGENT_KEY);

File: netty/src/main/java/io/grpc/netty/NettyClientStream.java (hunk 1 above)
Comment (zsurocking) [inline netty/src/main/java/io/grpc/netty/NettyClientStream.java]: Added a static variable to Utils.

**Comment by zsurocking** [comment]
PTAL. Thanks

**Comment by zsurocking** [comment]
Also, interop tests cannot be done at this moment. C server is the only one which supports GET, and the query string change is still pending.

File: netty/src/main/java/io/grpc/netty/NettyClientStream.java (hunk 1 above)
Comment (carl-mastrangelo) [inline netty/src/main/java/io/grpc/netty/NettyClientStream.java]: I am not sure this is done.

Comment (zsurocking) [inline netty/src/main/java/io/grpc/netty/NettyClientStream.java]: Actually done.

File: core/src/main/java/io/grpc/internal/AbstractClientStream2.java [hunk 9]
```diff
@@ -54,6 +56,15 @@
    * collisions/confusion. Only called from application thread.
    */
   protected interface Sink {
+    /** 
+     * Sends the request headers to the remote end point.
+     *
+     * @param metadata the metadata to be sent
+     * @param the payload needs to be sent in the headers if not null. Should only be used when
```
Comment (lukaszx0) [inline core/src/main/java/io/grpc/internal/AbstractClientStream2.java]: nit: `@param payload the payload needs to...`

File: core/src/main/java/io/grpc/internal/AbstractClientStream2.java [hunk 10]
```diff
@@ -166,6 +190,7 @@ public final boolean isReady() {
     private Runnable deliveryStalledTask;
 
     private boolean headersReceived;
+
```
Comment (lukaszx0) [inline core/src/main/java/io/grpc/internal/AbstractClientStream2.java]: ✂️

File: core/src/main/java/io/grpc/internal/Framer.java [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.internal;
+
+import io.grpc.Compressor;
+import java.io.InputStream;
+
+/** Interface for framing gRPC messages. */
+public interface Framer {
```
Comment (lukaszx0) [inline core/src/main/java/io/grpc/internal/Framer.java]: Either I missed something or it looks like it can be pkg private?

File: core/src/main/java/io/grpc/internal/AbstractClientStream2.java [hunk 12]
```diff
@@ -308,4 +333,69 @@ private void closeListener(Status status, Metadata trailers) {
...
+    @Override
+    public void dispose() {
+      closed = true;
+      payload = null;
+      headers = null;
+    }
+
+    @Override
+    public Framer setMessageCompression(boolean enable) {
+      return this;
```
Comment (lukaszx0) [inline core/src/main/java/io/grpc/internal/AbstractClientStream2.java]: Why is this not implemented? Will it ever be implemented? In any event, I'd suggest comment (TODO or explanation why it's unimplemented)

File: core/src/main/java/io/grpc/internal/AbstractClientStream2.java (hunk 9 above)
Comment (zsurocking) [inline core/src/main/java/io/grpc/internal/AbstractClientStream2.java]: Done

File: core/src/main/java/io/grpc/internal/AbstractClientStream2.java (hunk 10 above)
Comment (zsurocking) [inline core/src/main/java/io/grpc/internal/AbstractClientStream2.java]: Done

File: core/src/main/java/io/grpc/internal/Framer.java (hunk 11 above)
Comment (zsurocking) [inline core/src/main/java/io/grpc/internal/Framer.java]: Removed public

File: core/src/main/java/io/grpc/internal/AbstractClientStream2.java (hunk 12 above)
Comment (zsurocking) [inline core/src/main/java/io/grpc/internal/AbstractClientStream2.java]: Comments added.

File: core/src/main/java/io/grpc/internal/AbstractClientStream2.java [hunk 13]
```diff
@@ -384,6 +384,7 @@ public void dispose() {
       headers = null;
     }
 
+    // This framer sends everything in the header. So the methods below do not apply.
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/AbstractClientStream2.java]: It being in a header seems unrelated.

Comment (zsurocking) [inline core/src/main/java/io/grpc/internal/AbstractClientStream2.java]: We do call writeHeaders API only so I think it's related? Any suggestions on rephrasing it?

Comment (ejona86) [inline core/src/main/java/io/grpc/internal/AbstractClientStream2.java]: What does writeHeaders have to do with whether this does compression?

For compression, you could say that the GET encoding we have chosen doesn't support compression.

For max outbound message size, it seems maybe add a TODO?

Comment (zsurocking) [inline core/src/main/java/io/grpc/internal/AbstractClientStream2.java]: Done.

**Comment by ejona86** [comment]
Please make sure to squash.
